chore: upgrade `flair` to 0.12

alvarobartt opened the pull request:
# Description

As older versions of `flair` (< 0.12) were having problems during the installation, I've decided to upgrade `flair` to 0.12 and pin it. Besides the package upgrade, some things had to be updated in the `FlairMonitor` for the unit tests to pass when installing the latest version of `flair`.

Some more information about the installation issue at https://github.com/flairNLP/flair/issues/3129

**Type of change**

- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

- [x] Updated the `FlairMonitoring` unit tests

**Checklist**

- [X] I have merged the original branch into my forked branch
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works

alvarobartt committed:
Pin `flair` dependency to 0.12

alvarobartt committed:
Add `get_spans` to `FlairMonitor` due to upgrade

alvarobartt committed:
Put `flair` types under `typing.TYPE_CHECKING`

alvarobartt committed:
Fix `test_flair_monitoring` unit tests

alvarobartt committed:
Upgrade `flair` to 0.12 in `environment_dev.yml`

alvarobartt committed:
Revert `TYPE_CHECKING` usage in `FlairMonitor`

Breaks the unit tests because those imports previously included under `typing.TYPE_CHECKING` are imported from other module/s

frascuchon commented on src/argilla/monitoring/_flair.py:
```diff
@@ -63,7 +65,7 @@ def predict(self, sentences: Union[List[Sentence], Sentence], *args, **kwargs):
         result = self.__model__.predict(sentences, *args, **kwargs)
 
-        if isinstance(sentences, Sentence):
+        if not isinstance(sentences, list):
```
Be aware of using this condition only will match for `list`. If you're using a generator or other kinds of sequences to provide a sentence batch, the `predict` wrapper will fail. Even if the original method (`model.predict`) is working

```python
instance((i for i in [1,2,3]), list) # False
```

alvarobartt replied:
Even though you're right, I was following the type-hint for the `sentences` arg, that's why I kept that check as `isinstance(..., list)`, as I guess we're only expecting either a list of `Sentence` or a single `Sentence`.

alvarobartt committed:
Use `inspect.getfullargspec` for args & defaults

Use `inspect.getfullargspec` instead of the former approach which was more complex for no reason, since `getfullargspec` retrieves all the args and their defaults, if applicable

alvarobartt committed:
`shutil.rmtree` just if file exists in advance

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

alvarobartt committed:
Add missing return type-hints

alvarobartt committed:
Combine `_from_pretrained` & `from_pretrained` args

alvarobartt committed:
Update just existing `setfit` args in `update_config`

alvarobartt committed:
Apply suggestions from code review

Co-authored-by: David Berenstein <david.m.berenstein@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

alvarobartt committed:
Fix function-args retrieval and file-deletion after train (#3015)

# Description

Some `setfit` unit tests were not passing as it can be seen at
https://github.com/argilla-io/argilla/actions/runs/5113755403/jobs/9193313124,
with the exception `FileNotFound` which was due to the function defined
in `tests/training/helpers.py` named `train_with_cleanup` was not
properly cleaning the file, since it was trying to `shutil.rmtree` over
a file that could or could not exist. I assume that issue was not
appearing on other runs since probably the file was cached at some
point.

Then after solving that issue, another issue appeared `TypeError: got
multiple values for argument 'model_id'`, and `TypeError: got multiple
values for argument 'revision'`, which was also leading the `setfit`
tests to fail.

That was due to the function defined at `argilla/training/utils.py`
named `get_default_args` was not working as expected, since it was
having issues with the inheritance of the classmethod `from_pretrained`,
and adding some args that were not from the original function such as
`model_id` and `revision` which are args from the `_from_pretrained`
method, so on, since `from_pretrained` keeps the `kwargs` and then
propagates those to `_from_pretrained` (called as part of the return
statement), it was receiving the `model_id` and `revision` args set
during `from_pretrained`, as well as the ones propagated via `kwargs`
since those are not valid args in `from_pretrained`. So on, I've fixed
this by using `inspect.getfullargspec` instead, and that seems to work
without any issue when inheriting from multiple classes or even
inheriting from a MixIn.

**Type of change**

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] Improvement (change adding some improvement to an existing
functionality)

**How Has This Been Tested**

- [X] Re-run the `setfit` unit tests on Python 3.8, 3.9, and 3.10 and
all those are passing properly now

**Checklist**

- [X] I have merged the original branch into my forked branch
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [X] My changes generate no new warnings

gabrielmbmb reviewed (approved):
LGTM!

davidberenstein1957 reviewed (approved):
LGTM, but maybe the SetFit updates were supposed to go to another PR?
